### PR TITLE
Refactor `PredictionLoop.on_run_start` for consistency

### LINKED ISSUE #12730: [RFC] Discrepancies in [evaluation/prediction]_loop
## Proposed refactor

- [x] 1. There is a subtle difference in the way that methods are created within the evaluation loop and prediction loop as below. 

- [on_run_start in prediction_loop.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loops/dataloader/prediction_loop.py#L82-L84)
- [on_run_start in evaluation_loop.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loops/dataloader/evaluation_loop.py#L123-L135)

- [x] 2. The hook is not called through function (only in prediction_loop.py). 

- [_on_predict_model_eval in prediction_loop.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loops/dataloader/prediction_loop.py#L155-L158)
- As reference [_on_evaluation_model_eval in evaluation_loop.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loops/dataloader/evaluation_loop.py#L255-L260)

- [x] 3. The hooks called for testing and validation are different, and the way the variables are set at that time slightly different. 

- [_evaluation_epoch_end in evaluation_loop.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loops/dataloader/evaluation_loop.py#L296-L309)
- [_on_evaluation_epoch_end in evaluation_loop.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loops/dataloader/evaluation_loop.py#L311-L319)

### Motivation

1. For readability, it may be better to make the prediction loop methods with the same granularity as the evaluation method.

2. For consistency and readability

3. For consistency and readability

### Pitch

1. As below.

#### Before
```python
def on_run_start(self) -> None:  # type: ignore[override]
    """Calls ``_on_predict_start`` hook."""
    self._on_predict_start()
```

#### After
```python
def on_run_start(self) -> None:  # type: ignore[override]
    """Calls the ``_on_predict_model_eval``, ``_on_predict_start`` and ``_on_predict_epoch_start``
    hooks."""
    self._on_predict_model_eval()
    self.trainer.lightning_module.zero_grad()
    self._on_predict_start()
    self._on_predict_epoch_start()
```

2. As below.

#### Before
```python
def _on_predict_model_eval(self) -> None:
    """Calls ``on_predict_model_eval`` hook."""
    model_ref = self.trainer.lightning_module
    model_ref.on_predict_model_eval()
```

#### After
```python
def _on_predict_model_eval(self) -> None:
    """Calls ``on_predict_model_eval`` hook."""
    model_ref._call_lightning_module_hook("on_predict_model_eval")
```

3. As below. 

#### Before
```python
if self.trainer.testing:
    self.trainer._call_callback_hooks("on_test_epoch_start", *args, **kwargs)
    self.trainer._call_lightning_module_hook("on_test_epoch_start", *args, **kwargs)
else:
    self.trainer._call_callback_hooks("on_validation_epoch_start", *args, **kwargs)
    self.trainer._call_lightning_module_hook("on_validation_epoch_start", *args, **kwargs)
```

#### After
```python
stage = "test" if self.trainer.testing or "validation"
self.trainer._call_callback_hooks(f"on_{stage}_epoch_start", *args, **kwargs)
self.trainer._call_lightning_module_hook(f"on_{stage}_epoch_start", *args, **kwargs)
```

### Additional context

The impact is very limited. 
If this refactor makes sense, I will work on this.
______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @carmocca @justusschock @ananthsub @ninginthecloud @rohitgr7

### PR DESCRIPTION
**Pull request by ti-ginkgo**
## What does this PR do?

Fixes #12730

working on 1st issue.

> There is a subtle difference in the way that methods are created within the evaluation loop and prediction loop as below.
[on_run_start in prediction_loop.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loops/dataloader/prediction_loop.py#L82-L84)
[on_run_start in evaluation_loop.py](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loops/dataloader/evaluation_loop.py#L123-L135)

### Does your PR introduce any breaking changes? If yes, please list them.

No

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by carmocca — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]

**Comment by awaelchli** [comment]
@ti-ginkgo Yes, ignore HPU. 
Thanks for sending this refactor!

**Review by akihironitta — APPROVED** [review]
@ti-ginkgo LGTM! 🚀
